src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +32,134 @@ case class NoOpFingerprint() {
+
+    val rootFiles: Seq[FileInfo] =
+      leafDirToChildrenFiles.getOrElse(rootPath, Array()).map(FileInfo(_))
+
+    Content(Directory(rootPath.toString, rootFiles, subDirs))
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  // Create leaf subdirectory. Leaf Subdirectory is the first parent of leaf files.
+  private def createSubDirectoryTree(
+      dirPath: Path,
+      statuses: Array[FileStatus]): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = statuses.map(FileInfo(_)))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = statuses.map(FileInfo(_))))
+    }
+  }
+
+  // Create Non-Leaf subdirectory tree starting from the topmost non-root directory.
+  private def createNonLeafSubDirectoryHelper(dirPath: Path, child: Directory): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = Seq(), subDirs = Seq(child))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = Seq(), subDirs = Seq(child)))
     }
   }
 }
 
+case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+  // Create a Directory object from a directory path by recursively listing its leaf files. All
+  // files from the directory tree will be part of the Content.
+  def apply(path: Path): Directory = {
+    val files: Seq[FileInfo] = {
+      try {
+        val fs = path.getFileSystem(new Configuration)
+
+        // Filter out files like _SUCCESS.
+        val statuses = fs.listStatus(path).filter(status => isDataPath(status.getPath))
+
+        // TODO: Implement support for recursive listing of files below and remove the assert.
+        assert(statuses.forall(!_.isDirectory))
+
+        statuses.map(FileInfo(_)).toSeq
+      } catch {
+        // FileNotFoundException is an expected exception if path doesn't exist. This will happen,
+        // for e.g. when creating an IndexLogEntry object just before creation of an index.
+        case _: FileNotFoundException => Seq()
+        case e: Throwable => throw e
+      }
+    }
+    if (path.isRoot) {
+      Directory(path.toString, files)
+    } else {
+      createFromChild(path.getParent, Directory(path.getName, files))
+    }
+  }
+
+  /* from org.apache.spark.sql.execution.datasources.PartitionAwareFileIndex. */
+  private def isDataPath(path: Path): Boolean = {
+    val name = path.getName
+    !((name.startsWith("_") && !name.contains("=")) || name.startsWith("."))
+  }
+
+  def createFromChild(path: Path, child: Directory): Directory = {
```

Comment:
`private`? Also, something like `prependDirectory` makes more sense?

Reply:
yes thanks ! I was struggling to come up with a name 😆